src/util/mpsc/batch.rs

```diff
@@ -0,0 +1,345 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+use crossbeam::sync::AtomicOption;
+use crossbeam_channel;
+pub use crossbeam_channel::{RecvError, RecvTimeoutError, SendError, TryRecvError, TrySendError};
+use futures::task::{self, Task};
+use futures::{Async, Poll, Stream};
+
+struct State {
+    recv_task: AtomicOption<Task>,
+    notify_size: usize,
+    // How many messages are sent without notify.
+    old_pending: AtomicUsize,
+    notifier_registered: AtomicBool,
+}
+
+impl State {
+    fn new(notify_size: usize) -> State {
```

Comment:
assert!(notify_size > 0);

Reply:
Already checked in `bounded` and `unbounded`.